File: src/utils/command/ReceiverStatusUpdate.ts
export type ReceiverStatus = "open" | "close" | "error" | "connecting";

export type ReceiverStatusUpdateCmd = "receiverStatusUpdate";

export type ReceiverStatusUpdate = {
  cmd: ReceiverStatusUpdateCmd;
  data: {
    status: ReceiverStatus;
  };
};

export function getStatusUpdateMessage(
  status: ReceiverStatus
): ReceiverStatusUpdate {
  return {
    cmd: "receiverStatusUpdate",
    data: {
      status: status,
    },
  };
}

export function getStatusUpdateMessageCmd(): ReceiverStatusUpdateCmd {
  return "receiverStatusUpdate";
}

File: share/src/command_packet/app_command/receiver_status_update.rs
/*
 * Copyright 2021-2023 Golden_Water
 * SPDX-License-Identifier: AGPL-3.0-only
 */

#[derive(serde::Serialize, serde::Deserialize, PartialEq, Eq, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct ReceiverStatusUpdate {
  status: ReceiverStatus,
}

impl ReceiverStatusUpdate {
  pub fn new(status: ReceiverStatus) -> ReceiverStatusUpdate {
    ReceiverStatusUpdate { status }
  }
}

#[derive(serde::Serialize, serde::Deserialize, PartialEq, Eq, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub enum ReceiverStatus {
  Close,
  Connecting,
  Connected,

  Error,
  Reconnecting,

  Interrupted,
}
